File: testai-agent/cortex/reasoner.py
"""
TestAI Agent - Citation-Aware Reasoning Engine

The Cortex's reasoning layer that:
1. Queries the Brain for relevant knowledge
2. Tracks citations for zero-hallucination
3. Uses the Gateway for LLM calls (with limits)
4. Generates test plans with full traceability

Design Philosophy:
- Every claim must be backed by a source
- Never hallucinate - cite or admit uncertainty
- Show your work (visible reasoning)
- Be humble about confidence

Zero-Hallucination Approach:
- All knowledge comes from Brain (ChromaDB)
- Every Brain chunk has a citation (section, tags)
- LLM augments but doesn't invent facts
- Output always includes "Source: X" references
"""

from dataclasses import dataclass, field
from typing import List, Dict, Optional, Any, Tuple
from enum import Enum
import asyncio
import os
import sys

# Add parent to path for imports
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from brain.vector_store import QABrain, SearchResult, KnowledgeChunk
from connectors.llm_gateway import LLMGateway, GatewayResponse, Citation, create_gateway
from cortex.confidence import ConfidenceScorer, ConfidenceResult, ConfidenceLevel
from cortex.decision_engine import DecisionEngine, DecisionContext, Decision, ActionType


class ReasoningPhase(Enum):
    """Phases of the reasoning process."""
    UNDERSTANDING = "understanding"  # Parse what user wants
    RETRIEVING = "retrieving"        # Query Brain for knowledge
    PLANNING = "planning"            # Plan the approach
    GENERATING = "generating"        # Generate tests/analysis
    VALIDATING = "validating"        # Check results
    EXPLAINING = "explaining"        # Format output


@dataclass
class RetrievedKnowledge:
    """Knowledge retrieved from Brain with citations."""
    chunks: List[SearchResult]
    citations: List[Citation]
    total_relevance: float
    topics_covered: List[str]

    @property
    def has_knowledge(self) -> bool:
        return len(self.chunks) > 0

    @property
    def best_match(self) -> Optional[SearchResult]:
        return self.chunks[0] if self.chunks else None

    def format_for_prompt(self, max_chunks: int = 5) -> str:
        """Format knowledge for LLM prompt."""
        if not self.chunks:
            return "No specific knowledge found. Use general QA principles."

        lines = ["## Relevant Knowledge from QA Brain:\n"]
        for i, result in enumerate(self.chunks[:max_chunks], 1):
            chunk = result.chunk
            lines.append(f"### [{i}] {chunk.section}")
            lines.append(f"Tags: {', '.join(chunk.tags)}")
            lines.append(f"Relevance: {result.confidence:.0%}")
            lines.append(f"Content:\n{chunk.content[:500]}...")
            lines.append("")

        return "\n".join(lines)


@dataclass
class ReasoningResult:
    """Result of reasoning with full traceability."""
    phase: ReasoningPhase
    thinking: str              # Visible reasoning
    output: str               # The actual result
    citations: List[Citation]  # Sources used
    confidence: ConfidenceResult
    llm_response: Optional[GatewayResponse] = None
    knowledge_used: Optional[RetrievedKnowledge] = None

    @property
    def is_confident(self) -> bool:
        return self.confidence.can_proceed

    def format_with_sources(self) -> str:
        """Format output with sources appended."""
        result = self.output
        if self.citations:
            result += "\n\n---\n📚 **Sources:**\n"
            for i, citation in enumerate(self.citations, 1):
                conf_pct = int(citation.confidence * 100)
                result += f"  {i}. {citation.source} ({conf_pct}% match)\n"
                if citation.excerpt:
                    result += f"     > \"{citation.excerpt[:100]}...\"\n"
        return result


class Reasoner:
    """
    Citation-aware reasoning engine.

    Combines:
    - Brain (knowledge retrieval)
    - Gateway (LLM calls)
    - Decision Engine (confidence/actions)

    Usage:
        reasoner = Reasoner()

        # Reason about a feature
        result = await reasoner.reason_about_feature(
            feature="login page",
            user_request="Generate comprehensive tests"
        )

        print(result.format_with_sources())
    """

    # Expert QA prompts for different tasks
    PROMPTS = {
        "test_generation": '''You are Maya, a 12-year QA veteran. Generate test cases for this feature.

{knowledge}

FEATURE: {feature}
USER REQUEST: {request}
PAGE TYPE: {page_type}

Generate 5-7 specific, actionable test cases. For each test:
1. Clear title
2. Specific steps (not vague like "enter valid email" but "enter maya.test@company.com")
3. Expected result
4. Priority (Critical/High/Medium/Low)
5. Risk category (Security/Data/UX/Functionality)

Focus on edge cases and scenarios humans often miss.
Format as structured test cases.''',

        "security_analysis": '''You are a security-focused QA expert. Analyze this feature for vulnerabilities.

{knowledge}

FEATURE: {feature}
PAGE TYPE: {page_type}
ELEMENTS: {elements}

Identify:
1. Potential security vulnerabilities (OWASP Top 10)
2. Input validation gaps
3. Authentication/authorization issues
4. Data exposure risks

Be specific about attack vectors and mitigation.''',

        "edge_case_detection": '''You are an edge case specialist. Find edge cases for this feature.

{knowledge}

FEATURE: {feature}
PAGE TYPE: {page_type}

Identify edge cases for:
1. Boundary values
2. Empty/null states
3. Concurrency issues
4. Error handling gaps
5. Browser/device variations
6. Network conditions

Be specific and actionable.''',

        "clarification": '''Based on the context, what clarifying questions should we ask?

FEATURE: {feature}
USER REQUEST: {request}
KNOWLEDGE GAPS: {gaps}

Generate 2-3 specific, helpful clarifying questions.
Focus on information that would significantly improve test coverage.'''
    }

    def __init__(
        self,
        brain: Optional[QABrain] = None,
        gateway: Optional[LLMGateway] = None,
        confidence_threshold: float = 0.70,
    ):
        """
        Initialize the reasoner.

        Args:
            brain: QA Brain for knowledge retrieval (creates default if None)
            gateway: LLM Gateway (creates default if None)
            confidence_threshold: Minimum confidence to proceed
        """
        self.brain = brain
        self.gateway = gateway or create_gateway()
        self.decision_engine = DecisionEngine(confidence_threshold=confidence_threshold)
        self.scorer = ConfidenceScorer(default_threshold=confidence_threshold)

        # Lazy load brain
        self._brain_loaded = False

    def _ensure_brain(self) -> QABrain:
        """Ensure brain is loaded."""
        if self.brain is None:
            self.brain = QABrain()
            self._brain_loaded = True
        return self.brain

    async def retrieve_knowledge(
        self,
        query: str,
        page_type: Optional[str] = None,
        n_results: int = 5,
    ) -> RetrievedKnowledge:
        """
        Retrieve relevant knowledge from Brain.

        Args:
            query: Search query
            page_type: Optional page type filter
            n_results: Number of results to retrieve

        Returns:
            RetrievedKnowledge with chunks and citations
        """
        brain = self._ensure_brain()

        # Search brain
        if page_type:
            results = brain.get_for_page_type(page_type, n_results=n_results)
        else:
            results = brain.search(query, n_results=n_results)

        # Build citations
        citations = []
        topics = set()
        total_relevance = 0.0

        for result in results:
            chunk = result.chunk
            citations.append(Citation(
                source=f"Brain: {chunk.section}",
                chunk_id=result.id,
                confidence=result.confidence,
                excerpt=chunk.content[:150] if chunk.content else "",
            ))
            topics.update(chunk.tags)
            total_relevance += result.confidence

        avg_relevance = total_relevance / len(results) if results else 0.0

        return RetrievedKnowledge(
            chunks=results,
            citations=citations,
            total_relevance=avg_relevance,
            topics_covered=list(topics),
        )

    async def reason_about_feature(
        self,
        feature: str,
        user_request: Optional[str] = None,
        page_type: Optional[str] = None,
        page_elements: Optional[List[Dict]] = None,
    ) -> ReasoningResult:
        """
        Reason about a feature to generate test plan.

        Args:
            feature: Feature description
            user_request: User's specific request
            page_type: Type of page (login, signup, etc.)
            page_elements: Elements found on page

        Returns:
            ReasoningResult with tests and citations
        """
        thinking_steps = []

        # Phase 1: Understanding
        thinking_steps.append("💭 Understanding the request...")

        # Phase 2: Retrieve knowledge
        thinking_steps.append("💭 Searching QA Brain for relevant knowledge...")
        knowledge = await self.retrieve_knowledge(
            query=f"{feature} {page_type or ''} testing",
            page_type=page_type,
            n_results=5,
        )

        if knowledge.has_knowledge:
            thinking_steps.append(f"💭 Found {len(knowledge.chunks)} relevant knowledge chunks")
            thinking_steps.append(f"💭 Topics covered: {', '.join(knowledge.topics_covered[:5])}")
        else:
            thinking_steps.append("💭 No specific knowledge found, using general QA principles")

        # Phase 3: Calculate confidence
        confidence = self.scorer.score_generation(
            feature=feature,
            context_available=bool(page_type or page_elements),
            knowledge_chunks=len(knowledge.chunks),
        )
        thinking_steps.append(f"💭 Confidence: {confidence.level.value} ({confidence.score:.0%})")

        # Phase 4: Check if we can proceed
        if not confidence.can_proceed and not knowledge.has_knowledge:
            return ReasoningResult(
                phase=ReasoningPhase.UNDERSTANDING,
                thinking="\n".join(thinking_steps),
                output=f"I need more information to generate tests. {confidence.reasoning}",
                citations=knowledge.citations,
                confidence=confidence,
                knowledge_used=knowledge,
            )

        # Phase 5: Generate via LLM
        thinking_steps.append("💭 Generating test cases...")

        prompt = self.PROMPTS["test_generation"].format(
            knowledge=knowledge.format_for_prompt(),
            feature=feature,
            request=user_request or "Generate comprehensive tests",
            page_type=page_type or "unknown",
        )

        system_prompt = """You are Maya, a 12-year QA veteran. You generate specific, actionable test cases.
Key principles:
- Use real test data, not placeholders (e.g., "maya.test@company.com" not "valid email")
- Focus on edge cases humans miss
- Prioritize security and data integrity
- Be specific about expected behavior
- Format tests clearly with steps, data, and expected results"""

        response = await self.gateway.complete(
            prompt=prompt,
            system=system_prompt,
            citations=knowledge.citations,
            temperature=0.6,
            max_tokens=4096,
        )

        thinking_steps.append("💭 Validating generated tests...")

        # Build final result
        return ReasoningResult(
            phase=ReasoningPhase.GENERATING,
            thinking="\n".join(thinking_steps),
            output=response.content,
            citations=knowledge.citations + response.citations,
            confidence=confidence,
            llm_response=response,
            knowledge_used=knowledge,
        )

    async def analyze_security(
        self,
        feature: str,
        page_type: Optional[str] = None,
        elements: Optional[List[Dict]] = None,
    ) -> ReasoningResult:
        """Analyze security aspects of a feature."""
        thinking_steps = ["💭 Starting security analysis..."]

        # Retrieve security-focused knowledge
        knowledge = await self.retrieve_knowledge(
            query=f"security vulnerabilities {feature} {page_type or ''}",
            n_results=5,
        )
        thinking_steps.append(f"💭 Found {len(knowledge.chunks)} security-related knowledge chunks")

        # Format elements for prompt
        elements_str = ""
        if elements:
            elements_str = "\n".join([
                f"- {e.get('tag', 'unknown')}: {e.get('type', '')} ({e.get('name', '')})"
                for e in elements[:10]
            ])

        prompt = self.PROMPTS["security_analysis"].format(
            knowledge=knowledge.format_for_prompt(),
            feature=feature,
            page_type=page_type or "unknown",
            elements=elements_str or "No elements provided",
        )

        response = await self.gateway.complete(
            prompt=prompt,
            system="You are a security-focused QA expert. Identify vulnerabilities and provide specific mitigation recommendations.",
            citations=knowledge.citations,
            temperature=0.4,
        )

        confidence = self.scorer.score_security_analysis(
            page_type=page_type or "unknown",
            has_auth_elements=any(e.get('type') in ['password', 'email'] for e in (elements or [])),
            has_input_elements=any(e.get('tag') in ['input', 'textarea'] for e in (elements or [])),
            knowledge_match=knowledge.total_relevance,
        )

        return ReasoningResult(
            phase=ReasoningPhase.GENERATING,
            thinking="\n".join(thinking_steps),
            output=response.content,
            citations=knowledge.citations + response.citations,
            confidence=confidence,
            llm_response=response,
            knowledge_used=knowledge,
        )

    async def find_edge_cases(
        self,
        feature: str,
        page_type: Optional[str] = None,
    ) -> ReasoningResult:
        """Find edge cases for a feature."""
        thinking_steps = ["💭 Looking for edge cases..."]

        knowledge = await self.retrieve_knowledge(
            query=f"edge cases boundary testing {feature}",
            n_results=5,
        )
        thinking_steps.append(f"💭 Retrieved {len(knowledge.chunks)} relevant patterns")

        prompt = self.PROMPTS["edge_case_detection"].format(
            knowledge=knowledge.format_for_prompt(),
            feature=feature,
            page_type=page_type or "unknown",
        )

        response = await self.gateway.complete(
            prompt=prompt,
            system="You are an edge case specialist. Find scenarios that typical testers miss.",
            citations=knowledge.citations,
            temperature=0.5,
        )

        confidence = self.scorer.score_generation(
            feature=feature,
            context_available=bool(page_type),
            knowledge_chunks=len(knowledge.chunks),
        )

        return ReasoningResult(
            phase=ReasoningPhase.GENERATING,
            thinking="\n".join(thinking_steps),
            output=response.content,
            citations=knowledge.citations,
            confidence=confidence,
            llm_response=response,
            knowledge_used=knowledge,
        )

    async def generate_clarifications(
        self,
        feature: str,
        user_request: Optional[str] = None,
        gaps: Optional[List[str]] = None,
    ) -> List[str]:
        """Generate clarifying questions."""
        gaps = gaps or ["page type", "test scope", "priority areas"]

        prompt = self.PROMPTS["clarification"].format(
            feature=feature,
            request=user_request or "General testing",
            gaps=", ".join(gaps),
        )

        response = await self.gateway.complete(
            prompt=prompt,
            system="Generate helpful clarifying questions.",
            temperature=0.7,
            max_tokens=512,
        )

        # Parse questions from response
        lines = response.content.split("\n")
        questions = [
            line.strip().lstrip("123456789.-) ")
            for line in lines
            if line.strip() and "?" in line
        ]

        return questions[:3]

    def get_status(self) -> Dict[str, Any]:
        """Get reasoner status."""
        gateway_status = self.gateway.get_status()
        brain_status = {"loaded": self._brain_loaded}
        if self.brain:
            brain_status["chunk_count"] = len(self.brain._chunks) if hasattr(self.brain, '_chunks') else "unknown"

        return {
            "gateway": gateway_status,
            "brain": brain_status,
            "confidence_threshold": self.decision_engine.confidence_threshold,
        }

    def format_usage(self) -> str:
        """Get formatted usage status."""
        return self.gateway.format_usage_status()


# Convenience function
async def quick_reason(
    feature: str,
    page_type: Optional[str] = None,
) -> ReasoningResult:
    """Quick reasoning about a feature."""
    reasoner = Reasoner()
    return await reasoner.reason_about_feature(
        feature=feature,
        page_type=page_type,
    )


if __name__ == "__main__":
    async def main():
        print("🧠 TestAI Reasoner Demo")
        print("=" * 50)

        reasoner = Reasoner()
        print(reasoner.format_usage())
        print()

        # Test reasoning
        print("Testing: Login page analysis")
        result = await reasoner.reason_about_feature(
            feature="login form",
            page_type="login",
            user_request="Generate security-focused tests",
        )

        print("\n💭 Thinking:")
        print(result.thinking)
        print("\n📝 Result:")
        print(result.format_with_sources())

    asyncio.run(main())

File: electron/lib/cortex/reasoner.js
/**
 * Yali Agent - Citation-Aware Reasoning Engine
 * Ported from testai-agent/cortex/reasoner.py
 *
 * The Cortex's reasoning layer that:
 * 1. Queries the Brain for relevant knowledge
 * 2. Tracks citations for zero-hallucination
 * 3. Uses the Gateway for LLM calls
 * 4. Generates test plans with full traceability
 *
 * Design Philosophy:
 * - Every claim must be backed by a source
 * - Never hallucinate - cite or admit uncertainty
 * - Show your work (visible reasoning)
 * - Be humble about confidence
 *
 * Zero-Hallucination Approach:
 * - All knowledge comes from Brain (QA_BRAIN)
 * - Every Brain chunk has a citation (section, tags)
 * - LLM augments but doesn't invent facts
 * - Output always includes "Source: X" references
 */

const { ConfidenceScorer, ConfidenceLevel } = require('./confidence')
const { DecisionEngine, createDecisionContext, ActionType } = require('./decision-engine')
const { getForPageType, formatForPrompt, getEdgeCases } = require('./qa-brain')

/**
 * Reasoning phases
 */
const ReasoningPhase = {
  UNDERSTANDING: 'understanding',  // Parse what user wants
  RETRIEVING: 'retrieving',        // Query Brain for knowledge
  PLANNING: 'planning',            // Plan the approach
  GENERATING: 'generating',        // Generate tests/analysis
  VALIDATING: 'validating',        // Check results
  EXPLAINING: 'explaining'         // Format output
}

/**
 * Create retrieved knowledge object
 */
function createRetrievedKnowledge(sections = []) {
  const citations = sections.map(s => ({
    source: s.cite ? s.cite() : `Section ${s.id} - ${s.title}`,
    sectionId: s.id,
    confidence: 0.8, // Default confidence for brain knowledge
    excerpt: s.content ? s.content.slice(0, 150) : ''
  }))

  const topicsCovered = [...new Set(sections.flatMap(s => s.tags || []))]

  return {
    sections,
    citations,
    totalRelevance: sections.length > 0 ? 0.8 : 0,
    topicsCovered,

    get hasKnowledge() {
      return this.sections.length > 0
    },

    get bestMatch() {
      return this.sections[0] || null
    },

    formatForPrompt(maxChunks = 5) {
      return formatForPrompt(this.sections, maxChunks)
    }
  }
}

/**
 * Create reasoning result object
 */
function createReasoningResult(phase, thinking, output, citations, confidence, options = {}) {
  return {
    phase,
    thinking,               // Visible reasoning
    output,                 // The actual result
    citations,              // Sources used
    confidence,             // Confidence result
    llmResponse: options.llmResponse || null,
    knowledgeUsed: options.knowledgeUsed || null,

    get isConfident() {
      return this.confidence.canProceed
    },

    formatWithSources() {
      let result = this.output
      if (this.citations && this.citations.length > 0) {
        result += '\n\n---\n📚 **Sources:**\n'
        this.citations.forEach((citation, i) => {
          const confPct = Math.round((citation.confidence || 0.8) * 100)
          result += `  ${i + 1}. ${citation.source} (${confPct}% match)\n`
          if (citation.excerpt) {
            result += `     > "${citation.excerpt.slice(0, 100)}..."\n`
          }
        })
      }
      return result
    }
  }
}

/**
 * Expert QA prompts for different tasks
 */
const PROMPTS = {
  test_generation: `You are Yali, a 12-year QA veteran. Generate test cases for this feature.

{knowledge}

FEATURE: {feature}
USER REQUEST: {request}
PAGE TYPE: {page_type}

Generate 5-7 specific, actionable test cases. For each test:
1. Clear title
2. Specific steps (not vague like "enter valid email" but "enter yali.test@company.com")
3. Expected result
4. Priority (Critical/High/Medium/Low)
5. Risk category (Security/Data/UX/Functionality)

Focus on edge cases and scenarios humans often miss.
Format as structured test cases.`,

  security_analysis: `You are a security-focused QA expert. Analyze this feature for vulnerabilities.

{knowledge}

FEATURE: {feature}
PAGE TYPE: {page_type}
ELEMENTS: {elements}

Identify:
1. Potential security vulnerabilities (OWASP Top 10)
2. Input validation gaps
3. Authentication/authorization issues
4. Data exposure risks

Be specific about attack vectors and mitigation.`,

  edge_case_detection: `You are an edge case specialist. Find edge cases for this feature.

{knowledge}

FEATURE: {feature}
PAGE TYPE: {page_type}

Identify edge cases for:
1. Boundary values
2. Empty/null states
3. Concurrency issues
4. Error handling gaps
5. Browser/device variations
6. Network conditions

Be specific and actionable.`,

  clarification: `Based on the context, what clarifying questions should we ask?

FEATURE: {feature}
USER REQUEST: {request}
KNOWLEDGE GAPS: {gaps}

Generate 2-3 specific, helpful clarifying questions.
Focus on information that would significantly improve test coverage.`
}

/**
 * Reasoner class
 * Citation-aware reasoning engine
 */
class Reasoner {
  constructor(options = {}) {
    this.confidenceThreshold = options.confidenceThreshold || 0.70
    this.decisionEngine = new DecisionEngine(this.confidenceThreshold)
    this.scorer = new ConfidenceScorer(this.confidenceThreshold)
    this.callLLM = options.callLLM || null // External LLM function
  }

  /**
   * Retrieve relevant knowledge from Brain
   */
  async retrieveKnowledge(query, pageType = null, nResults = 5) {
    let sections = []

    if (pageType) {
      sections = getForPageType(pageType)
    } else {
      // Search by keywords in query
      const keywords = query.toLowerCase().split(/\s+/)
      const { searchByKeyword, searchByTag } = require('./qa-brain')

      for (const keyword of keywords) {
        const matches = searchByKeyword(keyword)
        sections.push(...matches)
      }

      // Deduplicate
      const seen = new Set()
      sections = sections.filter(s => {
        if (seen.has(s.id)) return false
        seen.add(s.id)
        return true
      })
    }

    // Limit results
    sections = sections.slice(0, nResults)

    return createRetrievedKnowledge(sections)
  }

  /**
   * Reason about a feature to generate test plan
   */
  async reasonAboutFeature(feature, userRequest = null, pageType = null, pageElements = []) {
    const thinkingSteps = []

    // Phase 1: Understanding
    thinkingSteps.push('💭 Understanding the request...')

    // Phase 2: Retrieve knowledge
    thinkingSteps.push('💭 Searching QA Brain for relevant knowledge...')
    const knowledge = await this.retrieveKnowledge(
      `${feature} ${pageType || ''} testing`,
      pageType,
      5
    )

    if (knowledge.hasKnowledge) {
      thinkingSteps.push(`💭 Found ${knowledge.sections.length} relevant knowledge chunks`)
      thinkingSteps.push(`💭 Topics covered: ${knowledge.topicsCovered.slice(0, 5).join(', ')}`)
    } else {
      thinkingSteps.push('💭 No specific knowledge found, using general QA principles')
    }

    // Phase 3: Calculate confidence
    const confidence = this.scorer.scoreGeneration(
      feature,
      Boolean(pageType || pageElements.length > 0),
      knowledge.sections.length
    )
    thinkingSteps.push(`💭 Confidence: ${confidence.level} (${Math.round(confidence.score * 100)}%)`)

    // Phase 4: Check if we can proceed
    if (!confidence.canProceed && !knowledge.hasKnowledge) {
      return createReasoningResult(
        ReasoningPhase.UNDERSTANDING,
        thinkingSteps.join('\n'),
        `I need more information to generate tests. ${confidence.reasoning}`,
        knowledge.citations,
        confidence,
        { knowledgeUsed: knowledge }
      )
    }

    // Phase 5: Generate via LLM (if available)
    thinkingSteps.push('💭 Generating test cases...')

    let output = ''
    let llmResponse = null

    if (this.callLLM) {
      const prompt = PROMPTS.test_generation
        .replace('{knowledge}', knowledge.formatForPrompt())
        .replace('{feature}', feature)
        .replace('{request}', userRequest || 'Generate comprehensive tests')
        .replace('{page_type}', pageType || 'unknown')

      const systemPrompt = `You are Yali, a 12-year QA veteran. You generate specific, actionable test cases.
Key principles:
- Use real test data, not placeholders (e.g., "yali.test@company.com" not "valid email")
- Focus on edge cases humans miss
- Prioritize security and data integrity
- Be specific about expected behavior
- Format tests clearly with steps, data, and expected results`

      try {
        llmResponse = await this.callLLM([
          { role: 'system', content: systemPrompt },
          { role: 'user', content: prompt }
        ], { maxTokens: 2000, temperature: 0.6 })
        output = llmResponse.content
      } catch (e) {
        output = this._generateFallbackTests(knowledge, pageType)
      }
    } else {
      output = this._generateFallbackTests(knowledge, pageType)
    }

    thinkingSteps.push('💭 Validating generated tests...')

    return createReasoningResult(
      ReasoningPhase.GENERATING,
      thinkingSteps.join('\n'),
      output,
      knowledge.citations,
      confidence,
      { llmResponse, knowledgeUsed: knowledge }
    )
  }

  /**
   * Analyze security aspects of a feature
   */
  async analyzeSecurity(feature, pageType = null, elements = []) {
    const thinkingSteps = ['💭 Starting security analysis...']

    // Retrieve security-focused knowledge
    const knowledge = await this.retrieveKnowledge(
      `security vulnerabilities ${feature} ${pageType || ''}`,
      pageType,
      5
    )
    thinkingSteps.push(`💭 Found ${knowledge.sections.length} security-related knowledge chunks`)

    // Format elements for prompt
    const elementsStr = elements.length > 0
      ? elements.slice(0, 10).map(e =>
          `- ${e.tag || 'unknown'}: ${e.type || ''} (${e.name || ''})`
        ).join('\n')
      : 'No elements provided'

    const hasAuthElements = elements.some(e =>
      ['password', 'email'].includes(e.type)
    )
    const hasInputElements = elements.some(e =>
      ['input', 'textarea'].includes(e.tag)
    )

    const confidence = this.scorer.scoreSecurityAnalysis(
      pageType || 'unknown',
      hasAuthElements,
      hasInputElements,
      knowledge.totalRelevance
    )

    let output = ''
    let llmResponse = null

    if (this.callLLM) {
      const prompt = PROMPTS.security_analysis
        .replace('{knowledge}', knowledge.formatForPrompt())
        .replace('{feature}', feature)
        .replace('{page_type}', pageType || 'unknown')
        .replace('{elements}', elementsStr)

      try {
        llmResponse = await this.callLLM([
          { role: 'system', content: 'You are a security-focused QA expert. Identify vulnerabilities and provide specific mitigation recommendations.' },
          { role: 'user', content: prompt }
        ], { maxTokens: 1500, temperature: 0.4 })
        output = llmResponse.content
      } catch (e) {
        output = this._generateFallbackSecurityAnalysis(knowledge, pageType)
      }
    } else {
      output = this._generateFallbackSecurityAnalysis(knowledge, pageType)
    }

    return createReasoningResult(
      ReasoningPhase.GENERATING,
      thinkingSteps.join('\n'),
      output,
      knowledge.citations,
      confidence,
      { llmResponse, knowledgeUsed: knowledge }
    )
  }

  /**
   * Find edge cases for a feature
   */
  async findEdgeCases(feature, pageType = null) {
    const thinkingSteps = ['💭 Looking for edge cases...']

    const knowledge = await this.retrieveKnowledge(
      `edge cases boundary testing ${feature}`,
      pageType,
      5
    )
    thinkingSteps.push(`💭 Retrieved ${knowledge.sections.length} relevant patterns`)

    const confidence = this.scorer.scoreGeneration(
      feature,
      Boolean(pageType),
      knowledge.sections.length
    )

    let output = ''
    let llmResponse = null

    if (this.callLLM) {
      const prompt = PROMPTS.edge_case_detection
        .replace('{knowledge}', knowledge.formatForPrompt())
        .replace('{feature}', feature)
        .replace('{page_type}', pageType || 'unknown')

      try {
        llmResponse = await this.callLLM([
          { role: 'system', content: 'You are an edge case specialist. Find scenarios that typical testers miss.' },
          { role: 'user', content: prompt }
        ], { maxTokens: 1500, temperature: 0.5 })
        output = llmResponse.content
      } catch (e) {
        output = this._generateFallbackEdgeCases(knowledge, pageType)
      }
    } else {
      output = this._generateFallbackEdgeCases(knowledge, pageType)
    }

    return createReasoningResult(
      ReasoningPhase.GENERATING,
      thinkingSteps.join('\n'),
      output,
      knowledge.citations,
      confidence,
      { llmResponse, knowledgeUsed: knowledge }
    )
  }

  /**
   * Generate clarifying questions
   */
  async generateClarifications(feature, userRequest = null, gaps = []) {
    const defaultGaps = gaps.length > 0 ? gaps : ['page type', 'test scope', 'priority areas']

    if (this.callLLM) {
      const prompt = PROMPTS.clarification
        .replace('{feature}', feature)
        .replace('{request}', userRequest || 'General testing')
        .replace('{gaps}', defaultGaps.join(', '))

      try {
        const response = await this.callLLM([
          { role: 'system', content: 'Generate helpful clarifying questions.' },
          { role: 'user', content: prompt }
        ], { maxTokens: 512, temperature: 0.7 })

        // Parse questions from response
        const lines = response.content.split('\n')
        const questions = lines
          .filter(line => line.trim() && line.includes('?'))
          .map(line => line.trim().replace(/^[\d.\-)\s]+/, ''))
          .slice(0, 3)

        return questions
      } catch (e) {
        // Fallback questions
      }
    }

    // Default clarifying questions
    return [
      `What type of page is this? (login, signup, checkout, etc.)`,
      `What are the most critical features to test?`,
      `Are there any known issues or areas of concern?`
    ]
  }

  /**
   * Fallback test generation when LLM is not available
   */
  _generateFallbackTests(knowledge, pageType) {
    const lines = ['## Generated Test Cases\n']

    if (knowledge.sections.length === 0) {
      lines.push('*No specific tests could be generated. Please provide more context.*')
      return lines.join('\n')
    }

    let tcNum = 1
    for (const section of knowledge.sections.slice(0, 3)) {
      lines.push(`### ${section.title}`)
      for (const test of section.tests.slice(0, 3)) {
        lines.push(`**TC-${String(tcNum).padStart(3, '0')}**: ${test.description}`)
        lines.push(`- Priority: ${test.priority}`)
        lines.push(`- ${section.cite()}`)
        lines.push('')
        tcNum++
      }
    }

    return lines.join('\n')
  }

  /**
   * Fallback security analysis
   */
  _generateFallbackSecurityAnalysis(knowledge, pageType) {
    const lines = ['## Security Analysis\n']

    const securitySections = knowledge.sections.filter(s =>
      s.tags.includes('security')
    )

    if (securitySections.length === 0) {
      lines.push('*Basic security checks recommended:*')
      lines.push('- Input validation on all fields')
      lines.push('- XSS prevention')
      lines.push('- CSRF token validation')
      lines.push('- Authentication/authorization checks')
      return lines.join('\n')
    }

    for (const section of securitySections) {
      lines.push(`### ${section.title}`)
      lines.push(section.content.slice(0, 300))
      lines.push('')
    }

    return lines.join('\n')
  }

  /**
   * Fallback edge case generation
   */
  _generateFallbackEdgeCases(knowledge, pageType) {
    const lines = ['## Edge Cases\n']

    // Get edge case data for common field types
    const edgeCases = {
      email: getEdgeCases('email'),
      password: getEdgeCases('password'),
      text: getEdgeCases('text'),
      number: getEdgeCases('number')
    }

    lines.push('### Input Field Edge Cases')
    lines.push('')
    lines.push('**Email Fields:**')
    edgeCases.email.slice(0, 5).forEach(ec => lines.push(`- ${ec}`))
    lines.push('')
    lines.push('**Text Fields:**')
    edgeCases.text.slice(0, 5).forEach(ec => lines.push(`- ${ec}`))
    lines.push('')
    lines.push('**Numeric Fields:**')
    edgeCases.number.slice(0, 5).forEach(ec => lines.push(`- ${ec}`))

    return lines.join('\n')
  }

  /**
   * Get reasoner status
   */
  getStatus() {
    return {
      confidenceThreshold: this.confidenceThreshold,
      hasLLM: Boolean(this.callLLM)
    }
  }
}

/**
 * Quick reasoning helper
 */
async function quickReason(feature, pageType = null, callLLM = null) {
  const reasoner = new Reasoner({ callLLM })
  return reasoner.reasonAboutFeature(feature, null, pageType)
}

module.exports = {
  ReasoningPhase,
  Reasoner,
  createRetrievedKnowledge,
  createReasoningResult,
  quickReason,
  PROMPTS
}
